R/settings.R

```diff
@@ -77,6 +78,72 @@ read_config_file <- function(config_file) {
   config
 }
 
+validate_config_file <- function(config, config_file, defaults) {
+  matched <- names(config) %in% names(defaults)
+  if (!all(matched)) {
+    warning("Found non-setting objects in config '", config_file, "': ", toString(names(config)[!matched]))
+  }
+
+  check_types <- intersect(names(config), names(defaults))
+  for (setting in check_types) {
+    observed_type <- typeof(config[[setting]])
+    expected_type <- typeof(defaults[[setting]])
+    if (observed_type != expected_type) {
+      stop("Setting '", setting, "' should be of type '", expected_type, "', not '", observed_type, "'.")
+    }
+  }
+
+  validate_linters(config)
+  validate_exclusions(config)
+}
+
+validate_linters <- function(config) {
```

Comment:
Don't we have a more advanced implementation of this already? e.g. auto-converting factories to linters by calling them.

Reply:
That's a deprecated feature though, right? So ideally we'd remove the code to do that.

Reply:
I searched for `"linter"` in our sources to find other related functionality, it turned up `is_linter()` as the `inherits(., "linter")` wrapper (which is useful here), but that's all I saw (possible I scanned past something useful).

Reply:
`validate_linter_object()`, or `define_linters()` seem useful.

In particular `linters: object_name_linter()` appears to be supported by the CLI usage (it's very handy indeed). Do we (currently) support providing a linter via settings instead of a list of linters?

Reply:
`validate_linter_object()`, or `define_linters()` seem useful.

In particular `linters: object_name_linter()` appears to be supported by the CLI usage (it's very handy indeed). Do we (currently) support providing a linter via settings instead of a list of linters?

Reply:
`validate_linter_object()`, or `define_linters()` seem useful.

In particular `linters: object_name_linter()` appears to be supported by the CLI usage (it's very handy indeed). Do we (currently) support providing a linter via settings instead of a list of linters?

Reply:
`validate_linter_object()`, or `define_linters()` seem useful.

In particular `linters: object_name_linter()` appears to be supported by the CLI usage (it's very handy indeed). Do we (currently) support providing a linter via settings instead of a list of linters?

Reply:
> In particular linters: object_name_linter() appears to be supported by the CLI usage

Is it? I tested at v3.1.0, current `main`, and this PR, all give an error for that:

```r
withr::local_dir(withr::local_tempdir())
writeLines("linters: infix_spaces_linter()", ".lintr")
writeLines("a=1", "aaa.R")
lint_dir()

# v3.1.0
# .Error in x[missing] : object of type 'closure' is not subsettable

# main @ 62858ba5675dbd4c86942bf1175dd605b97c4781
# Error in x[[1L]] : object of type 'closure' is not subsettable

# this branch @ 62858ba5675dbd4c86942bf1175dd605b97c4781
# Setting 'linters' should be of type 'list', not 'closure'.
```

Reply:
With CLI usage I meant the `linters=` argument for `lint()` and friends.
Since the answer is no, we don't support this shorthand for the config, I agree with your implementation 😊

Reply:
oh, got it. feel free to file as a follow-up FR, it seems reasonable enough & not too hard to support.

Reply:
#2230